Introducing Gazebo support for RX160/RX160L

MuriloMartins opened the pull request:
@gavanderhoorn Could you please verify this PR to add support for simulation of Staubli RX160/160L robot sin Gazebo 2.x.

It follows the design standards discussed in ros-industrial/abb_experimental#18.

Some remarks regarding this PR.
Firstly, I had to open the .stl files of some links in meshlab and save them again as .stl so that I could apply colour to those links in Gazebo (the original files were not working).

Also, I changed the colours to represent the real robots we have (they are actually yellow, not orange, but that is purely cosmetic).

Lastly, inertia matrices, as well as link masses, at the moment do not represent the actual real robot.

Although maximum speed and torque are documented in the robot manual, inertia matrices and/or COGs are not. I tried to follow [this tutorial](http://gazebosim.org/tutorials?tut=inertia&cat=build_robot) to extract the data from the CAD models, but I got matrices filled with nans for most joints.

I tested MoveIt with the simulated robot and it works just fine.

Commit:
Introducing Gazebo support for RX160/RX160L

gavanderhoorn commented:
Quick comment: there are numerous places that (re)introduce tabs instead of spaces, mostly in launch files and `xacro` macros. #14 changed those all to spaces. Could you make sure to use an indentation of 2 spaces for each level?

re: colours: the colour is based on information I found that Staubli used RAL 1028 for the RX160. If your robots have a different colour, please just add it to the relevant files. Removing the current xacro property it is not an option, as we do not know where it is being used.

MuriloMartins commented:
Thanks for pointing that out! I didn't realise vim was changing the indentation...

I fixed and rebased the PR.

I reverted the colour for the RX160, but left the new yellow to be used with the RX160L (I think that also helps distinguishing amongst them -- does that make sense?).

gavanderhoorn commented:
@MuriloMartins wrote:

> Thanks for pointing that out! I didn't realise vim was changing the indentation...
> 
> I fixed and rebased the PR.

ok, thanks for that.

> I reverted the colour for the RX160, but left the new yellow to be used with the RX160L (I think that also helps distinguishing amongst them -- does that make sense?).

Yes, this is ok.

I'm actually considering to just name the colours after their RAL name (ie: _Melon yellow_ for 1028, and _Zinc yellow_ for 1018). They already have a name, and it would make things more maintainable as well, instead of having to come up with names ourselves all the time. We could still have `colour_staubli_yellow`, but it would just be an alias of whatever it turns out that Staubli is actually using.

gavanderhoorn commented:
As to this PR: I hate to ask this of you, but could you split it into two? One against _this_ repository, containing just the changes to `resources` and `rx160_support`, and one against [ros-industrial/staubli_experimental](https://github.com/ros-industrial/staubli_experimental), introducing the experimental Gazebo pkg.

Nothing in the Staubli repositories has been released, but the `_experimental` repositories allow us to do some, well, _experimenting_ without worrying about backwards compatibility and support. I try to take that serious, hence my request.

Let me know if that is possible for you (I'm willing to do it myself).

gavanderhoorn commented:
I'll wait with reviewing the changes until after you've updated the PR(s).

MuriloMartins commented:
I split the PR and updated the colour names to use standard RAL naming (the macros are still using the names previously used to avoid changing URDF files - perhaps we should propagate the changes?).

The `staubli_rx160_gazebo` package is now the ros-industrial/staubli_experimental#1.

gavanderhoorn commented on staubli_resources/urdf/common_colours.xacro:
```diff
@@ -2,9 +2,15 @@
 
   <!-- RAL 1028 -->
-  <xacro:property name="colour_staubli_yellow" value="1.0 0.608 0 1" />
+  <xacro:property name="melon_yellow" value="1.0 0.608 0.0 1.0" />
```
Please keep the `colour_` prefix. Normal xacro does not have the concept of namespaces (so it's easy to end up with name clashes) and it makes sense to explicitly name entities (ie: `colour_` for colours, `material_` for materials, etc).

gavanderhoorn replied:
Also: we cannot remove the `colour_staubli_yellow` (yet), as we do not know if/where it has been used. We should add a `colour_staubli_ral_melon_yellow` and an alias called `colour_staubli_yellow` (which has `value="${colour_staubli_ral_melon_yellow}"`).

gavanderhoorn replied:
Also: I've been changing the way in which we define these colours: instead of doing the `int`->`float` conversion externally, I let xacro do it for us (see [fanuc_resources](https://github.com/ros-industrial/fanuc/blob/a859053f5b601012d279e74a7d769c8b428640fc/fanuc_resources/urdf/common_colours.xacro#L12-L25)) fi).

This saves us work (we don't have to do the conversion), and it makes it easier to figure out which particular RGB values were used for the RAL colour (seeing as not all sites give the same values for the same RAL colour).

Ideally we'd also include a comment stating which particular site was used to source the RGB values (like [abb_resources/common_colours.xacro](https://github.com/ros-industrial/abb_experimental/blob/ffb631009d044d7a38e7de485e3e9e2e65b924c7/abb_resources/urdf/common_colours.xacro#L3-L4). You don't have to use that site, it is just an example).

gavanderhoorn commented on staubli_resources/urdf/common_colours.xacro:
```diff
@@ -2,9 +2,15 @@
+  <!-- RAL 7043 -->
+  <xacro:property name="traffic_grey_b" value="0.306 0.329 0.322 1.0" />
 
 </robot>
```
General comment: I've asked Staubli directly, and RAL1028 is the official 'staubli yellow'. It would seem RAL1018 is not. It is however perfectly possible to get a robot in RAL1018, as you can get them in just about any colour.

I have some more info on other colours and will add them once we get this merged.

MuriloMartins replied:
I lost track of all desired changes regarding colours... :-\

gavanderhoorn replied:
Not a problem. We'll disregard it for now and I'll update things after we merge this PR.

gavanderhoorn replied:
Could you tell me which site you used to convert to rgb?

MuriloMartins replied:
I actually calculated those values based on the information I found at http://www.ralcolor.com/ (i.e., `RGB (approx.)` column values divided by 255, assuming `alpha = 1` since colours are solid).

gavanderhoorn commented on staubli_resources/urdf/common_materials.xacro:
```diff
@@ -4,13 +4,25 @@
+      <color rgba="${traffic_grey_b}"/>
     </material>
   </xacro:macro>
 
```
Here we should just mimic the 'new' colour naming scheme, so `material_staubli_ral_melon_yellow`, etc.

gavanderhoorn commented on staubli_rx160_support/urdf/rx160_macro.xacro:
```diff
@@ -2,6 +2,14 @@
+  <property name="link_3_mass" value="20.0" />
+  <property name="link_4_mass" value="10.0" />
+  <property name="link_5_mass" value="1.0" />
+  <property name="link_6_mass" value="0.05" />
```
I'm a big supporter of using constants instead of magic numbers, but 1) these seem to be only used in the `inertial` tags, and then only once, and 2) their names are too generic.

Either remove them and use the values directly, or prefix them with something like `staubli_rx160_` (as they are a property of that particular robot).

Personally I'd just use the values directly and remove these properties.

MuriloMartins replied:
I was experimenting with mass values, so it was easier to keep them in one place. I got the values extracted from the CAD models using Solidworks now, so they should not change anymore, and therefore I removed those properties in my latest commit.

gavanderhoorn commented on staubli_rx160_support/urdf/rx160_macro.xacro:
```diff
@@ -18,6 +26,12 @@
+          <inertia ixx="0.001" ixy="0.0" ixz="0.0"
+                   iyy="0.001" iyz="0.0"
+                   izz="0.001"/>
+        </inertial>
```
We'll have to add a note to the package manifest telling users that the inertia elements in the xacro are 'fictional'' (like in [kuka_experimental/kuka_kr16_support/package.xml](https://github.com/ros-industrial/kuka_experimental/blob/11d8a061b85431e6647f949486858f1dddd9cb59/kuka_kr16_support/package.xml#L28)).

MuriloMartins replied:
I got the actual values extracted from the official Staubli CAD models using Solidworks. I tested/checked moments of inertia in Gazebo.

gavanderhoorn replied:
In my experience most CAD models don't include things like wiring and other such things that have an influence on mass and inertia. I agree that these values (probably) are better than the default ones, but we can't really claim they are the 'real' values, so I feel that adding the comment is still justified.

We could mention that these are probably as good as they are going to get though.

MuriloMartins replied:
I agree with you, but unless Staubli provides us with the real values, or someone takes apart the robots and get the values empirically, this is indeed as good as it gets.

I did not claim them to be the real values, just actual values extracted from official CAD models using well-known and respectable CAD software. ;-)

You are right, a disclaimer in the package description is a sensible approach. I'll amend to my last commit!

BTW, the hard work with Solidworks was courtesy of @psotiropoulos.

gavanderhoorn replied:
> BTW, the hard work with Solidworks was courtesy of @psotiropoulos.

Much appreciated, thanks.

gavanderhoorn commented on staubli_rx160_support/urdf/rx160_macro.xacro:
```diff
@@ -117,42 +167,48 @@
       <child link="${prefix}link_1"/>
       <axis xyz="0 0 1"/>
-      <limit lower="-2.967060" upper="2.967060" effort="0" velocity="2.88" />
+      <limit lower="-2.967060" upper="2.967060" effort="692.0" velocity="2.88" />
```
Could you point me to the doc/section/page where you got these `effort` limits from?

MuriloMartins replied:
I have an older version of the Staubli instruction manual than the one referred to in `package.xml` (D28086504A), though I expect torque limits to be the same. In version D28075404C, the values can be found in a table in Section 2.6.2 (Torque Limits).

Note that I would allow maximum torque on axis 6, thus setting `effort` to 29 on axis 5, rather than a feasible 58 when torque on axis 6 = 0 (see superscripts (1) and (2) for details).

gavanderhoorn commented on staubli_rx160_support/urdf/rx160l_macro.xacro:
```diff
@@ -2,6 +2,14 @@
+  <property name="link_3_mass" value="1.0" />
+  <property name="link_4_mass" value="1.0" />
+  <property name="link_5_mass" value="1.0" />
+  <property name="link_6_mass" value="1.0" />
```
See earlier comment.

gavanderhoorn commented on staubli_rx160_support/urdf/rx160l_macro.xacro:
```diff
@@ -10,7 +18,7 @@
           <mesh filename="package://staubli_rx160_support/meshes/rx160/visual/base_link.stl"/>
         </geometry>
-        <xacro:material_staubli_yellow />
+        <xacro:material_staubli_yellow_b />
```
I know you've mentioned that your robot is actually this colour (RAL1018), but I'd prefer all robots in the support pkgs to have their 'standard colours' (so RAL1028). Users with custom colours should then update a personal fork.

I understand this will mean more work for you, but if we start adding robots to support pkgs based on whatever paint job users happen to have, that will not work for the majority of our users.

gavanderhoorn commented:
Some comments, mostly about how to name things (and then mostly the colour properties).

> Firstly, I had to open the .stl files of some links in meshlab and save them again as .stl so that I could apply colour to those links in Gazebo (the original files were not working).

Hm, I don't quite understand why that would be necessary. I'll try and see if I can get Gazebo to reproduce the issue you are describing. Looking at the 'rendered diff' Github shows, it looks like a few faces have been changed in `link_1.stl`. The meshes have been edited with meshlab before, so I don't really get why things would be different this time.

gavanderhoorn commented:
As to the RAL colours: these are obviously not Staubli specific, but without a prefix playing poor man's namespace, I think we run the risk of nameclashes (with unusable xacro macros as a result). So for now I suggest we just use `colour_staubli_ral_` as a prefix for all of those.

MuriloMartins commented:
@gavanderhoorn the difference in the `.stl` files might have been due to a version of meshlab, or the settings whilst saving the mesh as `.stl`. I literally opened the files and exported to `.stl`, having not done anything else, it is surprising a few faces have changed.

I could not apply colour to the previous meshes, and I accidentally realised that when analysing how the meshes used for collision in Gazebo influenced the ROS effort controllers I used (before switching to position controllers).

We could get the `.stl` files from the official CAD model again (though I assume that's where they came from at first place).

MuriloMartins commented:
@gavanderhoorn could we please revisit the desired changes regarding colours, I lost track of what we've discussed previously (sorry about that!).

gavanderhoorn commented:
> @gavanderhoorn could we please revisit the desired changes regarding colours, I lost track of what we've discussed previously (sorry about that!).

Yes, no problem. I'm preparing a PR (agains your PR) right now.

gavanderhoorn commented:
> @gavanderhoorn could we please revisit the desired changes regarding colours, I lost track of what we've discussed previously (sorry about that!).

@MuriloMartins: I've just submitted ocadotechnology/staubli#1. Does that clear things up?

Sorry for being to persistent with this, but I feel it'll pay off later.

MuriloMartins commented:
@gavanderhoorn  thanks for the changes, I've just merged your PR. I agree, the earlier we get things sorted the better.

I'll propagate the changes to the colour macros into my packages.

gavanderhoorn commented:
@MuriloMartins: we can go ahead and merge this. Two minor things:
1. please add [this comment](https://github.com/ros-industrial/staubli/pull/15/files#r35654320) (wording doesn't need to be identical) about the max joint torque for `joint_6` as an `NB` in the package manifest, similar to the note about the values in the `inertial` tags. I'd like to make such 'design decisions' (if those exist for urdfs) as explicit as possible. Perhaps also a quick note in the xacro macro itself, as a line-comment above the `joint_6` limits declaration?
2. squash all commits in this PR into a single "rx160: prepare for Gazebo support." (or something)

I'll merge the PR then.

MuriloMartins commented:
@gavanderhoorn Great, I'm on it. Could you please confirm that the torque limits from version D28075404C are the same as in version D28086504A of the instruction manual (in case you have access to it), just so we avoid mentioning 2 different versions of the manual as sources of information.

gavanderhoorn commented:
> Could you please confirm that the torque limits from version D28075404C are the same as in version D28086504A of the instruction manual

Yes, D28086504A has the same values.

MuriloMartins commented:
Thanks! I've just rebased the PR. Let me know if it looks OK. :-)

gavanderhoorn commented:
Yes, this is ok.

+1.

gavanderhoorn commented:
Thanks for contributing @MuriloMartins, much appreciated.

MuriloMartins commented:
No worries. :-)

gavanderhoorn commented:
@MuriloMartins wrote:

> Firstly, I had to open the .stl files of some links in meshlab and save them again as .stl so that I could apply colour to those links in Gazebo (the original files were not working).

I think I figured this one out: meshlab by default (at least on my machine) has _Materialise Color Encoding_ checked in the _Choose Saving Options_ dialogue. This would seem to be a _Materialise Magics_ compatible type of saving STLs, which embeds (non-standard) colour info in the STL header (see [STL (file format) - Binary STL - Color in binary STL](https://en.wikipedia.org/wiki/STL_%28file_format%29#Color_in_binary_STL) on Wikipedia).

I haven't checked, but it could be that the STL loader Gazebo uses recognises this, and Gazebo then doesn't allow any other colours to be configured for the mesh (just as RViz ignores `material` tags on Collada meshes). Another option is that Gazebo's STL loader is just confused, does something strange with the end result being the same: issues with configuring colours for meshes.
